### PR TITLE
[Merged by Bors] - feat: port LinearAlgebra.Basic

### PR DESCRIPTION
**Pull request by 4e554c4c**
---

- [x] depends on: #2105 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/Basic.lean
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Basic.lean:2723]: Note that we can remove this if #18345 goes through first.

Comment (eric-wieser) [reply Mathlib/LinearAlgebra/Basic.lean:2745]: That PR is merged, so this _could_ be deleted

Comment (4e554c4c) [reply Mathlib/LinearAlgebra/Basic.lean]: done in [4bfe92c](https://github.com/leanprover-community/mathlib4/pull/1979/commits/4bfe92c44242d4272277abc381c99dcd20074989)

**Comment by jcommelin** [comment]
Blocked by lean4#2003 and the likes

**Comment by kbuzzard** [comment]
I am more optimistic (having looked at the first error): I think it's blocked by the fact that https://github.com/leanprover/lean4/pull/2093 hasn't made it to mathlib yet but in contrast to lean4#2003 this change has already made it to lean 4 master.

**Comment by jcommelin** [comment]
I can confirm that the first error goes away after merging the bump branch

**Comment by jcommelin** [comment]
This PR will compile once (some version of) #2105 is merged.

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#2105
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#2105~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/LinearAlgebra/Basic.lean
Comment (RemyDegenne) [inline Mathlib/LinearAlgebra/Basic.lean:1382]:
```suggestion
    (f : M₂ →ₛₗ[τ₂₁] M) (hf) :
    range (codRestrict p f hf) = comap p.subtype (LinearMap.range f) := by
```

**Comment by Komyyy** [comment]
As a temporary remedy, this command removes all red errors.
```
attribute [-instance] Ring.toNonAssocRing
```

File: Mathlib/LinearAlgebra/Basic.lean
Comment (Komyyy) [inline Mathlib/LinearAlgebra/Basic.lean:68]: TODO: Erase this line.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**
